### PR TITLE
Add a codec AES_128_GCM_SIV for encrypting columns on disk

### PR DESCRIPTION
**Pull request by depressed-pho**
While this is implemented as a compression codec, it does not actually compress data. It instead encrypts data on disk. The key is obtained by executing a user-specified command at the server startup, or if it's not specified the codec refuses to process any data.

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Added a compression codec `AES_128_GCM_SIV` which encrypts columns instead of compressing them.

Detailed description / Documentation draft:
- Included in the commit.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

**Comment by alex-zaitsev** [comment]
@depressed-pho , this is nice, but can we have it more generic? I.e. instead of AES codec define it as Encrypted with parameters. That will allow to make it more generic, as current 'encrypt' function: https://clickhouse.tech/docs/en/sql-reference/functions/encryption-functions/

For now it will only support AES_128_GCM_SIV, but we can extend later to other options.

**Comment by depressed-pho** [comment]
Like `Codec(Default, Encrypted('AES-128-GCM-SIV'))`? Okay, I will do it.

**Comment by alexey-milovidov** [comment]
It should be conditionally compiled under `#if USE_SSL`, see examples in code.

**Comment by Enmk** [comment]
@depressed-pho thanks, a very nice feature! I noticed that there are no tests, could you please add some? To simplify this, you may want to reuse the unit-test suite from other codes, please see [`CodecTestCompatibility`](https://github.com/ClickHouse/ClickHouse/blob/ac989f3a8fe8e2135e25ac7d6e340b048092e906/src/Compression/tests/gtest_compressionCodec.cpp#L545) for example.

**Comment by alexey-milovidov** [comment]
Unit tests are optional but functional tests are mandatory.

**Comment by depressed-pho** [comment]
But I'm not sure how to write a functional test for this. Would it be enough to just add something like `<encryption><key_command>echo "Some static key in base64"</key_command></encryption>` to `tests/server-test.xml` and a stateless SQL test using the codec?

**Comment by alexey-milovidov** [comment]
@depressed-pho Yes, you can add a new config, see `tests/config/README.md`

**Comment by vzakaznikov** [comment]
If we make it more generic, it would nice for the encryption parameters to the codec to match our current `encrypt` function (parameter names, order and values).

**Comment by depressed-pho** [comment]
Updated my patch. It's now named `Encrypted(cipher)` where `cipher` has to be a literal `'AES-128-GCM-SIV'` for now. The code is guareded by `#if USE_SSL`, and has a stateless SQL test.

**Comment by depressed-pho** [comment]
Amended it again to fix a build failure occuring when !USE_SSL.

**Comment by depressed-pho** [comment]
Fixed !USE_SSL again.

**Comment by depressed-pho** [comment]
Marked the test as being skipped in fasttest, where SSL is disabled.

**Comment by depressed-pho** [comment]
Fixed several build errors found by the build check.

**Comment by depressed-pho** [comment]
And now only the Yandex-exclusive checks are failing. I can't do anything about that.

**Comment by depressed-pho** [comment]
Rebased and resolved a conflict.

**Comment by nikitamikhaylov** [comment]
@depressed-pho Please, run a script `generate-ya-make` from `utils` folder to solve an issue with `Yandex synchronization check`. Then commit add all ya.make files to this PR.

**Comment by depressed-pho** [comment]
> @depressed-pho Please, run a script `generate-ya-make` from `utils` folder to solve an issue with `Yandex synchronization check`. Then commit add all ya.make files to this PR.

I did, but the script updated no `ya.make` files except for the one already included in this commit. I rebased the commit anyway.

**Comment by alexey-milovidov** [comment]
Maybe we can make an interface to pass IV to codecs.
The IV should be pseudorandom but deterministic, like a hash from offset in file
(though two files will still have the same sequence of IV).

**Comment by alexey-milovidov** [comment]
This codec is actually mostly safe to use but some considerations exist:
- the information that is revealed is when some blocks of data are the same: it makes it unsafe if ClickHouse is used for multiple tenants and some of them may have access to encrypted data on disk - then one tenant can find that other tenant have the same sequences of data;
- also the size of compressed data is revealed if it is applied after compression - it makes bruteforce attacks slightly easier.

I think we can merge it and highlight these considerations in docs.

**Comment by alexey-milovidov** [comment]
@alex-zaitsev 

> For now it will only support AES_128_GCM_SIV, but we can extend later to other options.

For other modes of operation it will be "instant death" if we will not generate and store random IVs. But if we do it, replicas will argue about mismatched checksum of "compressed" data and download data parts from each other excessively.

So, I think we should not allow more modes of operation here.

**Review by FArthur-cmd — APPROVED** [review]

File: docs/en/operations/server-configuration-parameters/settings.md [hunk 1]
```diff
@@ -69,6 +69,28 @@ If no conditions met for a data part, ClickHouse uses the `lz4` compression.
...
+## encryption {#server-settings-encryption}
+
+Configures a command to obtain a key to be used by [encryption codecs](../../sql-reference/statements/create/table.md#create-query-encryption-codecs). The command, or a shell script, is expected to write a Base64-encoded key of any length to the stdout.
+
+**Example**
+
+For Linux with systemd:
+
+```xml
+<encryption>
```
Comment (alexey-milovidov) [inline docs/en/operations/server-configuration-parameters/settings.md:81]: This configuration can be confused with the configuration of entrypted disks.

File: programs/server/Server.cpp [hunk 2]
```diff
@@ -445,6 +455,39 @@ void checkForUsersNotInMainConfig(
...
+
+    auto process = ShellCommand::execute(key_command);
+
+    std::string b64_key;
+    readStringUntilEOF(b64_key, process->out);
+    process->wait();
+
+    // turbob64 doesn't like whitespace characters in input. Strip
+    // them before decoding.
+    std::erase_if(b64_key, [](char c)
```
Comment (alexey-milovidov) [inline programs/server/Server.cpp:470]: It has an inconsistency: in this place, the encryption key is in base64 while for encrypted disks it is in hex or binary.

Comment (alexey-milovidov) [reply programs/server/Server.cpp:470]: And I think, it's better to read it in hex.

File: programs/server/Server.cpp [hunk 3]
```diff
@@ -445,6 +455,39 @@ void checkForUsersNotInMainConfig(
     }
 }
 
+static void loadEncryptionKey(const std::string & key_command [[maybe_unused]], Poco::Logger * log)
+{
+#if USE_BASE64 && USE_SSL && USE_INTERNAL_SSL_LIBRARY
+
+    auto process = ShellCommand::execute(key_command);
```
Comment (alexey-milovidov) [inline programs/server/Server.cpp:462]: Another inconsistency - here the key is loaded by a command placed in config,
while for encrypted disks we don't allow to put a script in config file, we expect the key can be imported from env variable with `from_env=...` attribute in config.

File: src/Compression/CompressionCodecEncrypted.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+        return std::string(result.data(), 16);
+    }
+
+    void CompressionCodecEncrypted::encrypt(const std::string_view & plaintext, char * ciphertext_and_tag)
+    {
+        // Fixed nonce. Yes this is unrecommended, but we have to live
+        // with it.
+        std::string_view nonce("\0\0\0\0\0\0\0\0\0\0\0\0", 12);
```
Comment (alexey-milovidov) [inline src/Compression/CompressionCodecEncrypted.cpp:133]: It is actually ok for AES_GCM_SIV?

**Comment by alexey-milovidov** [comment]
Need help from @vitlibar for further polishing.

File: docs/en/operations/server-configuration-parameters/settings.md (hunk 1 above)
Comment (vitlibar) [reply docs/en/operations/server-configuration-parameters/settings.md:81]: Yes, a longer path in xml would be better, for example

<codecs>
    <aes_128_gcm_siv>
        <key_hex>...</key_hex>
    </aes_128_gcm_siv>
</codecs>

**Comment by vitlibar** [comment]
My thoughts about that:

1) About syntax. I believe `Codec(AES-128-GCM-SIV)` would be better than `Codec(Encrypted('AES-128-GCM-SIV'))` because the first one is shorter and more consistent with other codecs. I mean for ordinary compression codec we don't write `Codec(Compression('ZSTD'))`, don't we? And I think if we later decide to support other encryption algorithm, for example `AES-128-GCM` with storing random nonce in output stream, we will just create a new codec.

2) I suppose we should support multiple keys, because sometimes we may wish to switch to a new key but still allow using previous keys for decryption. Of course to support multiple keys we should add a key id into an output stream so the decryptor later will know which key to use even because in the future the current key can be changed.

3) As already said, the representation of keys in the config file looks confusing because we also have encrypted disks.
I suggest to set keys in the config file like this:

```
<encryption_codecs>
    <!-- just key in hex, not some command -->
    <!-- the code should use directly this key and throw an exception if its length is not 16 bytes -->
    <key_hex>...</key_hex>
</encryption_codecs>
```

or, for multiple keys

```
<encryption_codecs>
        <!-- just key in hex, not some command -->
        <!-- the code should use this key and throw an exception if its length is not 16 bytes -->
        <key_hex id="0">...</key_hex>
        <key_hex id="1">...</key_hex>
        <current_key_id>1</current_key_id>
</encryption_codecs>
```

or

```
<encryption_codecs>
        <!-- just key in hex, not some command -->
        <!-- the code should use this key and throw an exception if its length is not 16 bytes -->
        <key_hex id="0">...</key_hex>
        <key_hex id="1">...</key_hex>
    <aes_128_gcm_siv>
        <current_key_id>1</current_key_id>
    </aes_128_gcm_siv>
</encryption_codecs>
```

4) Also we should change the documentation and say there that though encryption keys can be stored in the configuration file that's is NOT recommended because it's not secure enough. It's better either to use some environmental variables (see "from_env" tag) or move the keys into a separate config file on a secure disk and put a symlink to that config file to `config.d/` folder.

5) About algorithm. It seems to be wrong. It says that is uses `aes_128_gcm_siv` but actually uses `aes_128_gcm`. Also I don't think we should use nonces containing only zeroes, a random hardcoded nonce would look better. With a corresponding comment about nonce misuse resistance. And if we require the keys in the config file to be 16 bytes long it seems to be unnecessary to call `HKDF`.

**Comment by alexey-milovidov** [comment]
> I believe Codec(AES-128-GCM-SIV) would be better than Codec(Encrypted('AES-128-GCM-SIV'))

Yes, 100%. It was just some false advise...

**Comment by alexey-milovidov** [comment]
> I suppose we should support multiple keys, because sometimes we may wish to switch to a new key but still allow using previous keys for decryption. Of course to support multiple keys we should add a key id into an output stream so the decryptor later will know which key to use in case if in the future the current key is changed.

+1, it will be consistent with full disk encryption that we have implemented.

**Comment by alexey-milovidov** [comment]
> And if we require the keys in the config file to be 16 bytes long it seems to be unnecessary to call HKDF.

+1.

Using KDF will make it inconsistent with full disk encryption.
Otherwise we can add a separate alternative way to specify key, when KDF will be applied.

**Pull request merged**